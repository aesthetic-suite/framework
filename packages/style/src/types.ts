import {
  ClassName,
  CSS,
  Direction,
  GenericProperties,
  Properties,
  Property,
  Unit,
  UnitFactory,
  Value,
  Variables,
} from '@aesthetic/types';

export type SheetType = 'global' | 'standard' | 'conditions';

export type Condition = string;

export type RankCache = Record<string, number>;

export interface ProcessOptions {
  deterministic?: boolean;
  direction?: Direction;
  rankings?: RankCache;
  type?: SheetType;
  unit?: Unit | UnitFactory;
  vendor?: boolean;
}

export interface RenderOptions extends ProcessOptions {
  className?: ClassName;
  conditions?: Condition[] | string;
  // minimumRank?: number;
  selector?: string;
}

export interface CacheItem {
  className: string;
  rank?: number;
}

export interface StyleRule {
  conditionText?: string;
  cssRules: StyleRule[];
  cssText: CSS;
  cssVariables: Variables<string>;
  textContent: CSS;
  type: number;
  insertRule: (rule: CSS, index: number) => number;
}

// ADDONS

export interface CacheManager {
  read: (key: string, minimumRank?: number) => CacheItem | null;
  write: (key: string, item: CacheItem) => void;
}

export interface DirectionConverter {
  convert: <T extends Value>(
    from: Direction,
    to: Direction,
    property: string,
    value: T,
  ) => { property: string; value: T };
}

export interface SheetManager {
  insertRule: (type: SheetType, rule: CSS, index?: number) => number;
}

export interface VendorPrefixer {
  prefix: (key: string, value: Value) => GenericProperties;
  prefixSelector: (selector: string, rule: CSS) => CSS;
}

export interface API {
  converter?: DirectionConverter | null;
  direction: Direction;
  prefixer?: VendorPrefixer | null;
}

export interface EngineOptions {
  cacheManager: CacheManager;
  direction: Direction;
  directionConverter?: DirectionConverter | null;
  sheetManager: SheetManager;
  vendorPrefixer?: VendorPrefixer | null;
}

export interface Engine {
  renderDeclaration: <K extends Property>(
    property: K,
    value: Properties[K],
    options?: RenderOptions,
  ) => ClassName;
}
