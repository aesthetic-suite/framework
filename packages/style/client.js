// Temporary!
module.exports = require('./lib/index.js');
