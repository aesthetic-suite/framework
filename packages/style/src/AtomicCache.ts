import { CacheItem, StyleParams } from './types';

// https://jsperf.com/javascript-objects-vs-map-performance
// https://jsperf.com/performance-of-map-vs-object
export default class AtomicCache {
  cache: { [property: string]: { [value: string]: CacheItem[] } } = {};

  match(item: CacheItem, params: StyleParams, minimumRank?: number): boolean {
    if (minimumRank !== undefined && item.rank < minimumRank) {
      return false;
    }

    if (item.selector !== params.selector) {
      return false;
    }

    if (item.type !== params.type) {
      return false;
    }

    if (Array.isArray(params.conditions) && params.conditions.length > 0) {
      return params.conditions.every(i =>
        item.conditions.find(p => p.query === i.query && p.type === i.type),
      );
    }

    return true;
  }

  read(
    property: string,
    value: string,
    params: StyleParams,
    minimumRank?: number,
  ): CacheItem | null {
    const propertyCache = this.cache[property];

    if (!propertyCache) {
      return null;
    }

    const valueCache = propertyCache[value];

    if (!valueCache || valueCache.length === 0) {
      return null;
    }

    return valueCache.find(item => this.match(item, params, minimumRank)) ?? null;
  }

  write(property: string, value: string, item: CacheItem): this {
    let propertyCache = this.cache[property];

    if (!propertyCache) {
      propertyCache = {};
      this.cache[property] = propertyCache;
    }

    let valueCache = propertyCache[value];

    if (!valueCache) {
      valueCache = [];
      propertyCache[value] = valueCache;
    }

    valueCache.push(item);

    return this;
  }
}
