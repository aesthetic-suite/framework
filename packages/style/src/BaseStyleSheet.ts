import { isSSR } from '@aesthetic/utils';
import { StyleRule } from './types';

export default abstract class BaseStyleSheet {
  lastIndex: number = -1;

  sheet: StyleRule;

  protected rafHandle: number = 0;

  protected ruleBuffer: { index: number; rule: string }[] = [];

  constructor(sheet: StyleRule) {
    this.sheet = sheet;
  }

  protected enqueueRule(rule: string, customIndex?: number): number {
    let index = 0;

    // When testing or SSR, just insert the rule immediately
    if (process.env.NODE_ENV === 'test' || isSSR()) {
      index = customIndex ?? this.sheet.cssRules.length;

      this.injectRule(rule, index);

      // Otherwise buffer the insert until the next animation frame
    } else {
      index = customIndex ?? (this.lastIndex += 1);

      this.ruleBuffer.push({
        index,
        rule,
      });

      if (!this.rafHandle) {
        this.rafHandle = requestAnimationFrame(this.flushRules);
      }
    }

    return index;
  }

  protected injectRule = (rule: string, index: number) => {
    // CSS is locked in developer tools when using `insertRule`,
    // so avoid using it in development so we have full control.
    // istanbul ignore next
    if (process.env.NODE_ENV === 'development' && !isSSR()) {
      this.sheet.textContent += rule;
    } else {
      this.sheet.insertRule(rule, index);
      this.lastIndex = this.sheet.cssRules.length;
    }
  };

  protected flushRules = () => {
    this.rafHandle = 0;
    this.ruleBuffer.forEach((buffer) => {
      this.injectRule(buffer.rule, buffer.index);
    });
  };
}
