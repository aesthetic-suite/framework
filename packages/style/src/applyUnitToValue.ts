import { isUnitlessProperty } from 'css-in-js-utils';
import { Value } from './types';

export default function applyUnitToValue(property: string, value: Value): string {
  if (typeof value === 'string' || isUnitlessProperty(property) || value === 0) {
    return String(value);
  }

  return `${value}px`;
}
