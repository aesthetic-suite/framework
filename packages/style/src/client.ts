import { Variables } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import createStyleElements, { getStyleElement } from './client/dom';
import hydrateStyles from './client/hydration';
// Rollup compatibility
import {
  createCacheManager,
  createStyleEngine,
  createSheetManager,
  formatVariable,
  EngineOptions,
  StyleEngine,
} from './index';

export { createStyleElements, getStyleElement };

function setRootVariables(vars: Variables) {
  const root = document.documentElement;

  objectLoop(vars, (value, key) => {
    root.style.setProperty(formatVariable(key), String(value));
  });
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  const engine: StyleEngine = {
    ...createStyleEngine({
      cacheManager: createCacheManager(),
      sheetManager: createSheetManager(createStyleElements()),
      ...options,
    }),
    setRootVariables,
  };

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
