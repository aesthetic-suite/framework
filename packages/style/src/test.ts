import {
	CacheManager,
	ClassName,
	CSS,
	EngineOptions,
	Sheet,
	SheetManager,
	SheetType,
} from '@aesthetic/types';
// Rollup compatibility
import {
	createCacheManager,
	createSheetManager,
	createStyleElements,
	createStyleEngine,
	getStyleElement,
	StyleEngine,
} from '.';

export function createTestCacheManager(): CacheManager<ClassName> {
	return createCacheManager();
}

export function createTestSheetManager(): SheetManager {
	return createSheetManager(createStyleElements());
}

export function createTestStyleEngine(
	options: Partial<EngineOptions<ClassName>> = {},
): StyleEngine {
	return createStyleEngine({
		cacheManager: createTestCacheManager(),
		sheetManager: createTestSheetManager(),
		...options,
	});
}

export function getRenderedStyles(type: Sheet | SheetType): CSS {
	return [...(typeof type === 'string' ? getStyleElement(type) : type).cssRules].reduce(
		(css, rule) => css + rule.cssText,
		'',
	);
}

export function purgeStyles(type?: SheetType): void {
	if (type) {
		// This is the only way to generate accurate snapshots.
		// It may slow down tests though?
		document.querySelector(`#aesthetic-${type}`)?.remove();
	} else {
		purgeStyles('global');
		purgeStyles('standard');
		purgeStyles('conditions');
	}
}
