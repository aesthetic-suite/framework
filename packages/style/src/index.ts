/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Direction, EngineOptions, VariablesMap } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { createStyleElements, getStyleElement } from './client/dom';
import { hydrateStyles } from './client/hydration';
import { createCacheManager } from './common/cache';
import { createStyleEngine } from './common/engine';
import { createSheetManager } from './common/sheet';
import { formatVariable } from './common/syntax';
import { StyleEngine } from './types';

export * from './common/cache';
export * from './common/constants';
export * from './common/engine';
export * from './common/helpers';
export * from './common/sheet';
export * from './common/syntax';
export * from './types';

export { createStyleElements, getStyleElement };

// Set active direction on the document `dir` attribute
function setDirection(direction: Direction) {
  document.documentElement.setAttribute('dir', direction);
}

// Set CSS variables to :root
function setRootVariables(vars: VariablesMap) {
  objectLoop(vars, (value, key) => {
    document.documentElement.style.setProperty(formatVariable(key), String(value));
  });
}

// Set active theme class names on the `body`
function setTheme(classNames: string[]) {
  document.body.className = classNames.join(' ');
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  if (global.AESTHETIC_CUSTOM_ENGINE) {
    return global.AESTHETIC_CUSTOM_ENGINE as StyleEngine;
  }

  const direction = (document.documentElement.getAttribute('dir') ||
    document.body.getAttribute('dir') ||
    'ltr') as Direction;

  const engine: StyleEngine = createStyleEngine({
    cacheManager: createCacheManager(),
    direction,
    sheetManager: createSheetManager(createStyleElements()),
    ...options,
  });

  // Match against browser preferences
  engine.prefersColorScheme = (scheme) => matchMedia(`(prefers-color-scheme: ${scheme})`).matches;
  engine.prefersContrastLevel = (level) => matchMedia(`(prefers-contrast: ${level})`).matches;

  // Handle DOM specific logic
  engine.setDirection = setDirection;
  engine.setRootVariables = setRootVariables;
  engine.setTheme = setTheme;

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
