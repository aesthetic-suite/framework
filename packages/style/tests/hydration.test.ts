import { createCacheManager } from '../src';
import { createClientEngine } from '../src/client';
import { CacheStorage, SheetType, StyleEngine } from '../src/types';
import { getRenderedStyles, purgeStyles } from '../src/test';

function createStyle(type: SheetType, lastIndex: number) {
  const style = document.createElement('style');

  style.type = 'text/css';
  style.media = 'screen';
  style.setAttribute('id', `aesthetic-${type}`);
  style.setAttribute('data-aesthetic-hydrate-index', String(lastIndex));
  style.setAttribute('data-aesthetic-rule-index', '21');
  style.setAttribute('data-aesthetic-type', type);

  document.head.append(style);

  return style;
}

describe('Hydration', () => {
  let cache: CacheStorage;
  let engine: StyleEngine;

  beforeEach(() => {
    const globe = createStyle('global', 2);
    globe.textContent = `:root { --font-size:16px;--bg-color:#fff;--fb-color:black; }@import url("test.css");@font-face { font-family:"Open Sans";font-style:normal;font-weight:800;src:url("fonts/OpenSans-Bold.woff2"); }@keyframes kf103rcyx { from { transform:translateX(0%); } to { transform:translateX(100%); }  }`;

    const standard = createStyle('standard', 16);
    standard.textContent = `.a { margin:0; }.b { padding:6px 12px; }.c { border:1px solid #2e6da4; }.d { border-radius:4px; }.e { display:inline-block; }.f { cursor:pointer; }.g { font-family:Roboto; }.h { font-weight:normal; }.i { line-height:normal; }.j { white-space:nowrap; }.k { text-decoration:none; }.l { text-align:left; }.m { background-color:#337ab7; }.n { vertical-align:middle; }.o { color:rgba(0, 0, 0, 0); }.p { animation-name:fade; }.q { animation-duration:.3s; }`;

    const conditions = createStyle('conditions', 3);
    conditions.textContent = `@media (width: 500px) { @media (width: 350px) { @supports (color: blue) { .t { color:blue; } } } }@media (width: 500px) { .s:hover { color:red; } }@media (width: 500px) { .r { margin:10px; } }@supports (color: green) { .u { color:green; } }`;

    cache = {};
    engine = createClientEngine({
      cacheManager: createCacheManager(cache),
    });
  });

  afterEach(() => {
    purgeStyles();
  });

  it('adds at-rules to the global cache', () => {
    expect(cache).toEqual(
      expect.objectContaining({
        '@import:test.css': [{ className: '' }],
        '@font-face:"Open Sans"': [{ className: '' }],
        '@keyframes:kf103rcyx': [{ className: '' }],
      }),
    );
    expect(engine.ruleIndex).toBe(21);
  });

  it('adds standard and condition rules to the class name cache', () => {
    expect(cache).toEqual(
      expect.objectContaining({
        'margin:0': [{ className: 'a', rank: 0 }],
        'padding:6px 12px': [{ className: 'b', rank: 1 }],
        'border:1px solid #2e6da4': [{ className: 'c', rank: 2 }],
        'border-radius:4px': [{ className: 'd', rank: 3 }],
        'display:inline-block': [{ className: 'e', rank: 4 }],
        'cursor:pointer': [{ className: 'f', rank: 5 }],
        'font-family:Roboto': [{ className: 'g', rank: 6 }],
        'font-weight:normal': [{ className: 'h', rank: 7 }],
        'line-height:normal': [{ className: 'i', rank: 8 }],
        'white-space:nowrap': [{ className: 'j', rank: 9 }],
        'text-decoration:none': [{ className: 'k', rank: 10 }],
        'text-align:left': [{ className: 'l', rank: 11 }],
        'background-color:#337ab7': [{ className: 'm', rank: 12 }],
        'vertical-align:middle': [{ className: 'n', rank: 13 }],
        'color:rgba(0, 0, 0, 0)': [{ className: 'o', rank: 14 }],
        'animation-name:fade': [{ className: 'p', rank: 15 }],
        'animation-duration:.3s': [{ className: 'q', rank: 16 }],
        'color:blue:@media (width: 500px):@media (width: 350px):@supports (color: blue)': [
          { className: 't', rank: 0 },
        ],
        'color:red::hover:@media (width: 500px)': [{ className: 's', rank: 1 }],
        'margin:10px:@media (width: 500px)': [{ className: 'r', rank: 2 }],
        'color:green:@supports (color: green)': [{ className: 'u', rank: 3 }],
      }),
    );
    expect(engine.ruleIndex).toBe(21);
  });

  it('doesnt re-insert hydrated font-face', () => {
    engine.renderFontFace({
      fontFamily: '"Open Sans"',
      fontStyle: 'normal',
      fontWeight: 800,
      src: 'url("fonts/OpenSans-Bold.woff2")',
    });

    expect(getRenderedStyles('global')).toMatchSnapshot();
  });

  it('doesnt re-insert hydrated import', () => {
    engine.renderImport('test.css');

    expect(getRenderedStyles('global')).toMatchSnapshot();
  });

  it('doesnt re-insert hydrated keyframes', () => {
    engine.renderKeyframes({
      from: { transform: 'translateX(0%)' },
      to: { transform: 'translateX(100%)' },
    });

    expect(getRenderedStyles('global')).toMatchSnapshot();
  });
});
