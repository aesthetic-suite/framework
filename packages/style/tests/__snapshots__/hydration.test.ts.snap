// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hydration doesnt re-insert hydrated font-face 1`] = `
":root {--font-size: 16px; --bg-color: #fff; --fb-color: black;}@import url(test.css);@font-face {font-family: \\"Open Sans\\"; font-style: normal; font-weight: 800; src: url(\\"fonts/OpenSans-Bold.woff2\\");}@keyframes kf103rcyx { 
  from {transform: translateX(0%);} 
  to {transform: translateX(100%);} 
}"
`;

exports[`Hydration doesnt re-insert hydrated import 1`] = `
":root {--font-size: 16px; --bg-color: #fff; --fb-color: black;}@import url(test.css);@font-face {font-family: \\"Open Sans\\"; font-style: normal; font-weight: 800; src: url(\\"fonts/OpenSans-Bold.woff2\\");}@keyframes kf103rcyx { 
  from {transform: translateX(0%);} 
  to {transform: translateX(100%);} 
}"
`;

exports[`Hydration doesnt re-insert hydrated keyframes 1`] = `
":root {--font-size: 16px; --bg-color: #fff; --fb-color: black;}@import url(test.css);@font-face {font-family: \\"Open Sans\\"; font-style: normal; font-weight: 800; src: url(\\"fonts/OpenSans-Bold.woff2\\");}@keyframes kf103rcyx { 
  from {transform: translateX(0%);} 
  to {transform: translateX(100%);} 
}"
`;
